src/pyqasm/pulse/openpulse_visitor.py

```diff
@@ -0,0 +1,379 @@
+                elif isinstance(wave_form_arg, pulse_ast.WaveformType):
+                    print("TODO")
+                    # evaluate expression
+                else:
+                    print("ERROR")
+
+            else:
+                print("ERROR")
+
+        ## capture command update waveform
+        if statement.name.name == "capture_v0 ":
+            print("TODO")
+        if statement.name.name == "capture_v1 ":
+            print("TODO")
+        if statement.name.name == "capture_v2 ":
+            print("TODO")
+        if statement.name.name == "capture_v3 ":
+            print("TODO")
+        if statement.name.name == "capture_v4 ":
+            print("TODO")
+
+        if self._check_only:
+            return 0, []
+
+        return 0, []
+
+    def _visit_delay_instruction(
+        self, statement: pulse_ast.DelayInstruction
+    ) -> list[pulse_ast.Statement]:
+        """Visit a delay instruction statement.
+
+        Args:
+            statement (pulse_ast.DelayInstruction): The delay instruction statement to visit.
+        Returns:
+            None
+        """
+        if isinstance(statement.duration, pulse_ast.DurationLiteral):
+            print("TODO")
+            # evaluate expression
+        elif isinstance(statement.duration, pulse_ast.DurationType):
+            print("TODO")
+            # evaluate expression
+        else:
+            print("ERROR")
+
+        if statement.qubits:
+            print("TODO")
+            # evaluate expression
+            # check if frame exists in global scope
+        else:
+            print("ERROR")
+
+        if self._check_only:
+            return []
+
+        return []
+
+    def visit_statement(self, statement: pulse_ast.Statement) -> list[pulse_ast.Statement]:
+        """Visit a statement element.
+
+        Args:
+            statement (pulse_ast.Statement): The statement to visit.
+
+        Returns:
+            None
+        """
+        logger.debug("Visiting statement '%s'", str(statement))
+        result = []
+        visit_map = {
+            pulse_ast.QuantumBarrier: self._visit_barrier,
+            pulse_ast.ClassicalDeclaration: self._visit_classical_declaration,
+            pulse_ast.ExpressionStatement: lambda x: self._visit_function_call(x.expression),
+            pulse_ast.DelayInstruction: self._visit_delay_instruction,
+        }
+
+        visitor_function = visit_map.get(type(statement))
+
+        if visitor_function:
+            if isinstance(statement, pulse_ast.ExpressionStatement):
+                # these return a tuple of return value and list of statements
+                _, ret_stmts = visitor_function(statement)  # type: ignore[operator]
+                result.extend(ret_stmts)
+            else:
+                result.extend(visitor_function(statement))  # type: ignore[operator]
+        else:
+            raise_qasm3_error(
+                f"Unsupported statement of type {type(statement)}",
```

Comment:
This will need to be clearer, that this is an unsupported statement inside the Pulse grammar. This is to ensure that users are not confused that, say, `QubitDeclaration`(I'm assuming those are not allowed in Pulse?) made inside the `cal` block is an invalid QASM statement. It should be evident from the error message that we are parsing the Pulse level description of a program and in the given context, this statement is invalid